builder/module-config.nix

```
@@ -0,0 +1,178 @@
+{ config, pkgs, lib, ... }:
+let
+  inherit (import ../files.nix {
+    inherit pkgs;
+    inherit (config) release;
+    inherit (config.hardware) target;
+    variant = config.hardware.subtarget;
+    packagesArch = config.hardware.arch;
+    feedsSha256 = config.feedsHash;
+    kmodsSha256 = config.kmodsHash;
+    sha256 = config.sumsFileHash;
+  }) variantFiles profiles expandDeps corePackages packagesByFeed allPackages;
+
+
+  customPackagesFormatted = config.packages.include ++ (builtins.map (pkg: "-${pkg}") config.packages.exclude);
+
+  requiredPackages =
+    profiles.default_packages or (builtins.attrNames packagesByFeed.base ++ builtins.attrNames corePackages) ++
+    profiles.profiles.${config.hardware.profile}.device_packages or [ ] ++
+    customPackagesFormatted;
+
+  allRequiredPackageFiles = expandDeps allPackages requiredPackages;
+
+  mkPackageSymlink = pname:
+    let
+      package = allPackages.${pname};
+    in
+    lib.optional (package.type == "real") ''
+      [ -e "packages/${package.filename}" ] || ln -s ${package.file} "packages/${package.filename}"
+    '';
+
+  symlinkPackages = pkgs.writeScript "symlink-openwrt-packages" (
+    lib.concatLines (
+      lib.foldl' (acc: pname: acc ++ (mkPackageSymlink pname)) [ ] allRequiredPackageFiles
+    )
+  );
+
+  mkCopyStatement = file: ''
+    (
+     target_rel=${lib.escapeShellArg file.target}
+     target="files/''${target_rel#/}"
+     ${lib.optionalString (lib.pathIsRegularFile file.source) ''
+        mkdir -p "$(dirname "''${target}")"
+     ''}
+     cp -r --no-preserve=all "${file.source}" "''${target}"
+    )
+  '';
+
+  copyFiles = pkgs.writeScript "copy-openwrt-files" (
+    lib.concatLines (builtins.map mkCopyStatement config.files)
+  );
```

I actually prefer having once-used definitions inline for readability. The indentation isn't very deep here so I think it would be fine:

```suggestion
  copyFiles = pkgs.writeScript "copy-openwrt-files" (
    lib.concatMapStrings (file: ''
      (
        target_rel=${lib.escapeShellArg file.target}
        target="files/''${target_rel#/}"
        ${lib.optionalString (lib.pathIsRegularFile file.source) ''
          mkdir -p "$(dirname "''${target}")"
        ''}
        cp -r --no-preserve=all "${file.source}" "''${target}"
      )
    '') config.files
  );
```